File: packages/schemas/spectra_schemas/investigation.py
"""Investigation state, agent trace and the structured answer contract."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from .enums import (
    AgentName,
    AnswerStatus,
    ClaimStatus,
    ConfidenceLabel,
    InvestigationStatus,
    Modality,
    QueryIntent,
    SearchMode,
    TraceStatus,
)
from .evidence import (
    ApplicationLink,
    Claim,
    Contradiction,
    EvidenceLedger,
    TimelineEvent,
    VerificationResult,
)


def _now() -> datetime:
    return datetime.now(timezone.utc)


class TraceStep(BaseModel):
    """One observable action.  Action summaries only - never model chain-of-thought."""

    model_config = ConfigDict(extra="forbid")

    step_id: str
    investigation_id: str
    sequence: int
    agent: AgentName
    tool: str | None = None
    status: TraceStatus = TraceStatus.STARTED
    title: str = ""
    input_summary: str = ""
    output_summary: str = ""
    started_at: datetime = Field(default_factory=_now)
    completed_at: datetime | None = None
    latency_ms: float = 0.0
    evidence_ids: list[str] = Field(default_factory=list)
    error: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class ToolInvocation(BaseModel):
    model_config = ConfigDict(extra="forbid")

    call_id: str
    tool: str
    arguments: dict[str, Any] = Field(default_factory=dict)
    started_at: datetime = Field(default_factory=_now)
    completed_at: datetime | None = None
    latency_ms: float = 0.0
    ok: bool = True
    error: str | None = None
    result_summary: str = ""
    result_count: int = 0


class BudgetState(BaseModel):
    model_config = ConfigDict(extra="forbid")

    mode: SearchMode = SearchMode.DEEP
    max_tool_calls: int = 30
    max_latency_seconds: float = 60.0
    max_iterations: int = 8
    tool_calls_used: int = 0
    iterations_used: int = 0
    elapsed_seconds: float = 0.0
    exhausted_reason: str | None = None

    @property
    def tool_calls_remaining(self) -> int:
        return max(self.max_tool_calls - self.tool_calls_used, 0)

    def is_exhausted(self) -> bool:
        return (
            self.tool_calls_used >= self.max_tool_calls
            or self.iterations_used >= self.max_iterations
            or self.elapsed_seconds >= self.max_latency_seconds
        )


class QueryUnderstanding(BaseModel):
    model_config = ConfigDict(extra="forbid")

    original: str
    rewritten: str = ""
    intent: QueryIntent = QueryIntent.SEMANTIC_SEARCH
    complexity: str = "simple"
    detected_ids: list[str] = Field(default_factory=list)
    target_modalities: list[Modality] = Field(default_factory=list)
    temporal_hint: str | None = None
    keywords: list[str] = Field(default_factory=list)
    explanation: str = ""
    produced_by: str = "rule-router"


class AgentAvailability(BaseModel):
    name: str
    enabled: bool
    ready: bool
    reason: str | None = None
    alternatives: list[str] = Field(default_factory=list)


class InvestigationMetrics(BaseModel):
    model_config = ConfigDict(extra="forbid", protected_namespaces=())

    total_latency_ms: float = 0.0
    tool_calls: int = 0
    iterations: int = 0
    candidates_retrieved: int = 0
    evidence_used: int = 0
    evidence_rejected: int = 0
    contradictions: int = 0
    claims_made: int = 0
    claims_refuted: int = 0
    model_latency_ms: dict[str, float] = Field(default_factory=dict)
    models_used: list[str] = Field(default_factory=list)
    stage_latency_ms: dict[str, float] = Field(default_factory=dict)
    gpu_peak_mb: float | None = None
    tokens: dict[str, int] = Field(default_factory=dict)
    sources_considered: list[str] = Field(default_factory=list)


class InvestigationState(BaseModel):
    """Serialisable state of one investigation - the agent's whole memory."""

    model_config = ConfigDict(extra="forbid")

    investigation_id: str
    case_id: str | None = None
    goal: str
    mode: SearchMode = SearchMode.DEEP
    status: InvestigationStatus = InvestigationStatus.CREATED
    understanding: QueryUnderstanding | None = None
    plan: list[str] = Field(default_factory=list)
    entities: list[str] = Field(default_factory=list)
    evidence: EvidenceLedger = Field(default_factory=EvidenceLedger)
    contradictions: list[Contradiction] = Field(default_factory=list)
    timeline: list[TimelineEvent] = Field(default_factory=list)
    verification: VerificationResult | None = None
    tool_history: list[ToolInvocation] = Field(default_factory=list)
    trace: list[TraceStep] = Field(default_factory=list)
    budget: BudgetState = Field(default_factory=BudgetState)
    confidence: float = 0.0
    answer: str = ""
    answer_status: AnswerStatus = AnswerStatus.INSUFFICIENT_EVIDENCE
    application_links: list[ApplicationLink] = Field(default_factory=list)
    claims: list[Claim] = Field(default_factory=list)
    metrics: InvestigationMetrics = Field(default_factory=InvestigationMetrics)
    degraded: bool = False
    degraded_reasons: list[str] = Field(default_factory=list)
    agent_availability: list[AgentAvailability] = Field(default_factory=list)
    uploaded_asset_ids: list[str] = Field(default_factory=list)
    followups: list[str] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=_now)
    updated_at: datetime = Field(default_factory=_now)
    user_id: str | None = None
    role: str = "analyst"

    @property
    def confidence_label(self) -> ConfidenceLabel:
        return ConfidenceLabel.from_score(self.confidence)

    def leading_claim(self) -> Claim | None:
        """The best-supported claim - what the answer is built around."""
        live = [c for c in self.claims if c.status is not ClaimStatus.REFUTED]
        return max(live, key=lambda c: c.confidence, default=None)


class QueryExplanation(BaseModel):
    """User-facing 'why SPECTRA searched these sources'."""

    reasons: list[str] = Field(default_factory=list)
    sources_selected: list[str] = Field(default_factory=list)
    sources_skipped: list[dict[str, str]] = Field(default_factory=list)


class SearchAutopsy(BaseModel):
    """Post-investigation forensics - the demonstration centrepiece."""

    model_config = ConfigDict(extra="forbid")

    investigation_id: str
    sources_considered: list[str] = Field(default_factory=list)
    candidates_retrieved: int = 0
    evidence_used: int = 0
    evidence_rejected: int = 0
    rejection_reasons: dict[str, int] = Field(default_factory=dict)
    tool_calls: int = 0
    tool_breakdown: dict[str, int] = Field(default_factory=dict)
    contradictions: int = 0
    total_latency_ms: float = 0.0
    stage_latency_ms: dict[str, float] = Field(default_factory=dict)
    models_used: list[str] = Field(default_factory=list)
    gpu_peak_mb: float | None = None
    claims_made: int = 0
    claims_refuted: int = 0
    evidence_diversity: float = 0.0
    degraded: bool = False
    degraded_reasons: list[str] = Field(default_factory=list)


class InvestigationAnswer(BaseModel):
    """The structured response contract.

    The application NEVER parses free-form model text for behaviour - it reads
    these fields.
    """

    model_config = ConfigDict(extra="forbid")

    investigation_id: str
    answer: str
    confidence: float
    confidence_label: ConfidenceLabel
    status: AnswerStatus
    entities: list[dict[str, Any]] = Field(default_factory=list)
    evidence: list[dict[str, Any]] = Field(default_factory=list)
    contradictions: list[Contradiction] = Field(default_factory=list)
    timeline: list[TimelineEvent] = Field(default_factory=list)
    claims: list[Claim] = Field(default_factory=list)
    application_links: list[ApplicationLink] = Field(default_factory=list)
    explanation: QueryExplanation = Field(default_factory=QueryExplanation)
    autopsy: SearchAutopsy | None = None
    metrics: InvestigationMetrics = Field(default_factory=InvestigationMetrics)
    degraded: bool = False
    degraded_reasons: list[str] = Field(default_factory=list)
    followups: list[str] = Field(default_factory=list)


class InvestigationCase(BaseModel):
    """Persistent case that can be reopened and continued."""

    model_config = ConfigDict(extra="forbid")

    case_id: str
    title: str
    question: str
    investigation_ids: list[str] = Field(default_factory=list)
    entity_ids: list[str] = Field(default_factory=list)
    status: InvestigationStatus = InvestigationStatus.CREATED
    created_at: datetime = Field(default_factory=_now)
    updated_at: datetime = Field(default_factory=_now)
    evidence_count: int = 0
    contradiction_count: int = 0
    claim_count: int = 0
    confidence: float = 0.0

File: apps/frontend/src/lib/schemas/investigation.ts
import { z } from 'zod';

import {
  agentNameSchema,
  answerStatusSchema,
  confidenceLabelSchema,
  investigationStatusSchema,
  isoDateTimeSchema,
  searchModeSchema,
  traceStatusSchema,
} from './primitives';
import {
  applicationLinkSchema,
  canonicalEntitySchema,
  claimSchema,
  contradictionSchema,
  evidenceItemSchema,
  timelineEventSchema,
} from './evidence';

export const traceStepSchema = z.object({
  step_id: z.string(),
  /** Absent on SSE `trace` frames, present on `GET /api/investigations/{id}/trace`. */
  investigation_id: z.string().optional(),
  sequence: z.number().int(),
  agent: agentNameSchema,
  tool: z.string().nullable().default(null),
  status: traceStatusSchema.default('started'),
  title: z.string().default(''),
  input_summary: z.string().default(''),
  output_summary: z.string().default(''),
  started_at: isoDateTimeSchema.optional(),
  completed_at: isoDateTimeSchema.nullable().optional(),
  latency_ms: z.number().default(0),
  evidence_ids: z.array(z.string()).default([]),
  error: z.string().nullable().default(null),
  metadata: z.record(z.unknown()).default({}),
});
export type TraceStep = z.infer<typeof traceStepSchema>;

export const budgetStateSchema = z.object({
  mode: searchModeSchema.optional(),
  max_tool_calls: z.number().default(30),
  max_latency_seconds: z.number().optional(),
  max_iterations: z.number().optional(),
  tool_calls_used: z.number().default(0),
  iterations_used: z.number().optional(),
  elapsed_seconds: z.number().optional(),
  exhausted_reason: z.string().nullable().optional(),
});
export type BudgetState = z.infer<typeof budgetStateSchema>;

export const investigationMetricsSchema = z.object({
  total_latency_ms: z.number().default(0),
  tool_calls: z.number().default(0),
  iterations: z.number().default(0),
  candidates_retrieved: z.number().default(0),
  evidence_used: z.number().default(0),
  evidence_rejected: z.number().default(0),
  contradictions: z.number().default(0),
  claims_made: z.number().default(0),
  claims_refuted: z.number().default(0),
  model_latency_ms: z.record(z.number()).default({}),
  models_used: z.array(z.string()).default([]),
  stage_latency_ms: z.record(z.number()).default({}),
  gpu_peak_mb: z.number().nullable().default(null),
  tokens: z.record(z.number()).default({}),
  sources_considered: z.array(z.string()).default([]),
});
export type InvestigationMetrics = z.infer<typeof investigationMetricsSchema>;

export const queryExplanationSchema = z.object({
  reasons: z.array(z.string()).default([]),
  sources_selected: z.array(z.string()).default([]),
  sources_skipped: z.array(z.record(z.string())).default([]),
});
export type QueryExplanation = z.infer<typeof queryExplanationSchema>;

export const searchAutopsySchema = z.object({
  investigation_id: z.string(),
  sources_considered: z.array(z.string()).default([]),
  candidates_retrieved: z.number().default(0),
  evidence_used: z.number().default(0),
  evidence_rejected: z.number().default(0),
  rejection_reasons: z.record(z.number()).default({}),
  tool_calls: z.number().default(0),
  tool_breakdown: z.record(z.number()).default({}),
  contradictions: z.number().default(0),
  total_latency_ms: z.number().default(0),
  stage_latency_ms: z.record(z.number()).default({}),
  models_used: z.array(z.string()).default([]),
  gpu_peak_mb: z.number().nullable().default(null),
  claims_made: z.number().default(0),
  claims_refuted: z.number().default(0),
  evidence_diversity: z.number().default(0),
  degraded: z.boolean().default(false),
  degraded_reasons: z.array(z.string()).default([]),
});
export type SearchAutopsy = z.infer<typeof searchAutopsySchema>;

/**
 * `InvestigationAnswer.entities` is typed `list[dict[str, Any]]` server-side and
 * documented as canonical-entity rows. It is parsed with the canonical shape but
 * tolerates partial rows, because the server type does not guarantee every field.
 */
const answerEntitySchema = canonicalEntitySchema.partial().extend({
  entity_id: z.string(),
  canonical_name: z.string().optional(),
});
export type AnswerEntity = z.infer<typeof answerEntitySchema>;

export const investigationAnswerSchema = z.object({
  investigation_id: z.string(),
  answer: z.string(),
  confidence: z.number(),
  confidence_label: confidenceLabelSchema,
  status: answerStatusSchema,
  entities: z.array(answerEntitySchema).default([]),
  evidence: z.array(evidenceItemSchema).default([]),
  contradictions: z.array(contradictionSchema).default([]),
  timeline: z.array(timelineEventSchema).default([]),
  claims: z.array(claimSchema).default([]),
  application_links: z.array(applicationLinkSchema).default([]),
  explanation: queryExplanationSchema.default({
    reasons: [],
    sources_selected: [],
    sources_skipped: [],
  }),
  autopsy: searchAutopsySchema.nullable().default(null),
  metrics: investigationMetricsSchema.default({
    total_latency_ms: 0,
    tool_calls: 0,
    iterations: 0,
    candidates_retrieved: 0,
    evidence_used: 0,
    evidence_rejected: 0,
    contradictions: 0,
    claims_made: 0,
    claims_refuted: 0,
    model_latency_ms: {},
    models_used: [],
    stage_latency_ms: {},
    gpu_peak_mb: null,
    tokens: {},
    sources_considered: [],
  }),
  degraded: z.boolean().default(false),
  degraded_reasons: z.array(z.string()).default([]),
  followups: z.array(z.string()).default([]),
});
export type InvestigationAnswer = z.infer<typeof investigationAnswerSchema>;

export const createInvestigationResponseSchema = z.object({
  investigation_id: z.string(),
  case_id: z.string().nullable().default(null),
  status: z.string(),
  stream_url: z.string(),
});
export type CreateInvestigationResponse = z.infer<typeof createInvestigationResponseSchema>;

/**
 * `GET /api/investigations` is documented only as "recent investigations (summary
 * rows)". The row shape is not specified, so everything beyond the id is optional
 * and the UI degrades field by field rather than refusing the list.
 */
export const investigationSummarySchema = z.object({
  investigation_id: z.string(),
  case_id: z.string().nullable().optional(),
  question: z.string().optional(),
  goal: z.string().optional(),
  status: investigationStatusSchema.optional(),
  answer_status: answerStatusSchema.optional(),
  mode: searchModeSchema.optional(),
  confidence: z.number().optional(),
  confidence_label: confidenceLabelSchema.optional(),
  evidence_count: z.number().optional(),
  contradiction_count: z.number().optional(),
  created_at: isoDateTimeSchema.optional(),
  updated_at: isoDateTimeSchema.optional(),
});
export type InvestigationSummary = z.infer<typeof investigationSummarySchema>;

export const investigationCaseSchema = z.object({
  case_id: z.string(),
  title: z.string(),
  question: z.string(),
  investigation_ids: z.array(z.string()).default([]),
  entity_ids: z.array(z.string()).default([]),
  status: investigationStatusSchema.default('created'),
  created_at: isoDateTimeSchema.optional(),
  updated_at: isoDateTimeSchema.optional(),
  evidence_count: z.number().default(0),
  contradiction_count: z.number().default(0),
  claim_count: z.number().default(0),
  confidence: z.number().default(0),
});
export type InvestigationCase = z.infer<typeof investigationCaseSchema>;

/* --------------------------------------------------------------------------
 * SSE frames — `GET /api/stream/investigation/{id}`
 * ----------------------------------------------------------------------- */

export const streamClaimsFrameSchema = z.object({
  claims: z.array(claimSchema).default([]),
});

export const streamEvidenceFrameSchema = z.object({
  evidence: z.array(evidenceItemSchema).default([]),
});

export const streamStatusFrameSchema = z.object({
  status: z.string(),
  iteration: z.number().optional(),
  budget: budgetStateSchema.optional(),
});
export type StreamStatusFrame = z.infer<typeof streamStatusFrameSchema>;

export const streamErrorFrameSchema = z.object({
  error: z.string(),
  recoverable: z.boolean().default(false),
});
export type StreamErrorFrame = z.infer<typeof streamErrorFrameSchema>;
